modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java

```diff
@@ -77,14 +75,13 @@ public ClientCachePartitionsRequest(BinaryRawReader reader) {
             if (cacheDesc == null)
                 continue;
 
-            ClientCachePartitionAwarenessGroup grp = processCache(ctx, groups, cacheGroupIds, affinityVer, cacheDesc);
+            ClientCachePartitionAwarenessGroup grp = processCache(ctx, affinityVer, cacheDesc);
 
-            // Cache already processed.
             if (grp == null)
                 continue;
 
-            groups.add(grp);
-            cacheGroupIds.put(cacheDesc.groupId(), grp);
+            grps.computeIfAbsent(grp, grp0 -> grp::addCache)
```

Comment:
Complex logic, hard to understand, moreover addCache called twice for the first cache. I think it's better to use something line `Map<ClientCachePartitionAwarenessGroup, ClientCachePartitionAwarenessGroup> grps` instead and add groups like:
```
ClientCachePartitionAwarenessGroup oldGrp = grps.putIfAbsent(grp, grp);
if (oldGrp != null)
  oldGrp.addCache(cacheDesc);
```